### PR TITLE
To do api

### PR DESCRIPTION
**Pull request by Ru-T**
## What Changed?
- Create controller actions for 4 endpoints: get index, post create, put update, delete destroy
- Create serializer for to_do
- Controller tests 
## Why?
- To enable ember app to get the responses it expects from its requests

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,56 +1,3 @@
...
-  #   resources :posts, concerns: :toggleable
-  #   resources :photos, concerns: :toggleable
-
-  # Example resource route within a namespace:
-  #   namespace :admin do
-  #     # Directs /admin/products/* to Admin::ProductsController
-  #     # (app/controllers/admin/products_controller.rb)
-  #     resources :products
-  #   end
+  resources :to_dos, only: [:index, :create, :update, :destroy]
```
Comment (albaer) [inline config/routes.rb:2]: Good use of `only` with `resources`! It's really easy to create more routes than you are actually using, so this is a really good habit.

File: app/models/to_do.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class ToDo < ActiveRecord::Base
+
+  def as_json
+    { id: id, title: title, is_completed: is_completed }
```
Comment (albaer) [inline app/models/to_do.rb]: This works. As something to think about, we usually use [ActiveModelSerializers](https://github.com/rails-api/active_model_serializers) for a more OO approach or at least [jbuilder](https://github.com/rails/jbuilder) to keep it out of the model.

File: spec/controllers/to_dos_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
+require 'rails_helper'
+
+RSpec.describe ToDosController, type: :controller do
+
+    let(:to_do) {FactoryGirl.create(:to_do)}
```
Comment (albaer) [inline spec/controllers/to_dos_controller_spec.rb]: I don't think you have a linter yet, but best practice is to put spaces inside a block, like: `{ FactoryGirl.create(:to_do) }`.

File: spec/controllers/to_dos_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    describe "GET /to_dos.json" do
+      before do
+        get :index
+      end
+      it "returns all to_dos in json format" do
+        expect(response.content_type).to eq "application/json"
+        expect(response).to include
```
Comment (albaer) [inline spec/controllers/to_dos_controller_spec.rb]: There are a couple of things that will make it easier to test a json api.

File: app/serializers/to_do_serializer.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class ToDoSerializer < ActiveModel::Serializer
+  attributes :id, :title, :is_completed
```
Comment (albaer) [inline app/serializers/to_do_serializer.rb:2]: Hooray, AMS!

File: spec/controllers/to_dos_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
+require 'rails_helper'
+
+RSpec.describe ToDosController, type: :controller do
+
+    let(:to_do) { FactoryGirl.create(:to_do) }
+    let(:json) { JSON.parse(response.body) }
```
Comment (albaer) [inline spec/controllers/to_dos_controller_spec.rb:6]: Very nice.

File: app/controllers/to_dos_controller.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    @to_do = ToDo.new(to_do_params)
+    if @to_do.save
+      render json: @to_do
+    else
+      render json: @to_do.errors
+    end
+  end
+
+  def update
+    @to_do.update(to_do_params)
```
Comment (iandonovan) [inline app/controllers/to_dos_controller.rb]: Small point: `update` will perform the save, so believe it or not you're saving the record twice here. You can just do `if @to_do.update(to_do_params)` and it'll do the same thing slightly more efficiently.

File: app/models/to_do.rb [hunk 8]
```diff
@@ -0,0 +1,3 @@
+class ToDo < ActiveRecord::Base
+
```
Comment (iandonovan) [inline app/models/to_do.rb:2]: I don't remember if the syllabus requires validations, but consider adding some! What if the todo has no title, for instance; is that valid data?

File: spec/controllers/to_dos_controller_spec.rb (hunk 6 above)
Comment (iandonovan) [inline spec/controllers/to_dos_controller_spec.rb:6]: :+1:

File: spec/controllers/to_dos_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        expect(json["to_do"]["is_completed"]).to eq false
+      end
+    end
+
+    describe "DELETE /to_dos/:id.json" do
+      before do
+        delete :destroy, id: to_do.id
+      end
+      it "deletes a to_do" do
+        expect(ToDo.find_by_id(to_do.id)).to be_nil
```
Comment (iandonovan) [inline spec/controllers/to_dos_controller_spec.rb]: This is good, but think about how you can use the [rspec change matcher](https://www.relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/change-matcher) to assert something about how the number of Todos might change. :)

File: spec/controllers/to_dos_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    describe "POST /to_dos.json" do
+      before do
+        post :create, to_do: { title: "New title", is_completed: true }
+      end
+      it "creates a new to_do" do
+        expect(ToDo.last.title).to eq "New title"
+        expect(json["to_do"]["title"]).to eq "New title"
+        expect(json["to_do"]["is_completed"]).to eq true
```
Comment (iandonovan) [inline spec/controllers/to_dos_controller_spec.rb:28]: Same thing here, actually; sorry for going a bit out of order. You can perhaps assert that the number of Todos in the DB changed by +1.

File: app/controllers/to_dos_controller.rb (hunk 7 above)
Comment (Ru-T) [inline app/controllers/to_dos_controller.rb]: Good call.

File: spec/controllers/to_dos_controller_spec.rb (hunk 10 above)
Comment (Ru-T) [inline spec/controllers/to_dos_controller_spec.rb:28]: Will do - thanks!

**Pull request merged**